Features/task 3

kubmir opened the pull request:
I think task 3 is done - it is merged with the current state of task 2 but task 2 is not approved yet. I do not know when I will be at work (maybe 2nd/3rd January or maybe not) so I created this PR and I suggest to take it as a pre-review if somebody will have time to review it - if not no problem.

Commit:
KA-228 Add ConnectionStrings.config

Commit:
KA-228 Add DatabaseConnectionLoader with interface

kubmir committed:
KA-227 Add empty ConnectionStrings for CI build

Commit:
KA-228 Connect repository gets/create methods to mongo

Commit:
KA-228 Refactor DatabaseConnectionLoader

Issue with connection string was caused by wrong name of xml attribute
in ConnectionStrings.config. Name according to article on the Internet is
value however application require name connectionString

Commit:
KA-228 Update .gitignore

Commit:
KA-228 Make classes internal

Make DatabaseConnectionLoader, UrlLocationHelper and NotesRepository
internal - they were public.

Commit:
KA-228 Extend note model with dates

Commit:
KA-228 Add interfaces for services

Commit:
KA-228 Extend NoteEqualityComparer - add dates to comparison

Commit:
KA-228 Add NotesServicesTests

Commit:
KA-228 Add DateService implementation

Commit:
KA-228 Add NotesServices implementation

Commit:
KA-228 Register types for DI of classes in services

Commit:
KA-228 Update controller tests to use service layer

Commit:
KA-228 Remove unnecessary repository dependency

Commit:
KA-228 Fix build issues with parsing of date

Commit:
KA-228 Add dummy service function for update and delete

Commit:
KA-228 Add DatabaseContext class with connectionString as DI

Commit:
Merge branch 'features/task-2' into features/task-3

Commit:
KA-228 Update DI container for connectionString as DI

Commit:
KA-228 Add tests for extended funcionality of controller

Commit:
KA-228 Project clean up

Commit:
KA-228 Add validation of note

Commit:
Merge branch 'features/task-2' into features/task-3

Commit:
KA-228 Modify registration of types to be compatible with task-3

Commit:
KA-228 Update gitignore for connection strings

Commit:
KA-228 Add JSON formatter

Commit:
KA-228 Add update and delete functionality

Commit:
KA-228 Simplify possible lambda functions in NotesServices

Commit:
KA-228 Add basic update and delete tests in NotesServices

Commit:
Merge branch 'features/task-2' into features/task-3

Commit:
KA-228 Refactor tests

Commit:
KA-228 Improve controller put and delete methods with tests

Commit:
KA-228 Change signature of update methods and tests

Commit:
KA-228 Add GuidService

Commit:
KA-228 Refactor service layer together with tests

Remove class NotesServices and divide its functionality into AddNoteService,
UpdateNoteService and NotesRepository. Service layer is implemented for
operations with some business logic - add, update. Due to refactoring of
services, controller is also refactored to use new add/update services and
repository for get/delete operations.

Commit:
KA-228 Add GetNoteService with caching together with tests

Commit:
KA-228 Refactor visibility of service classes and tests

Commit:
KA-228 Refactor controller to use getNoteService

Commit:
KA-228 Refactor using of ModelState

Commit:
Merge branch 'features/task-2' into features/task-3

Commit:
KA-228 Remove references to mongo from contracts

Commit:
KA-228 Refactor registration of connection string in DI container

Commit:
KA-228 Refactor RegisterDependency in DependencyContainerBuilder

Commit:
KA-228 Clean solution

Commit:
KA-228 Fix bug with update of note

PetrSvirak commented on NotesBackEnd/src/Notes.Api/App_Start/DependencyInjectionConfig.cs:
```diff
@@ -11,11 +12,14 @@ internal static class DependencyInjectionConfig
         {
-            IDependencyContainerResolver container = DependencyContainerBuilder.SetUpApiContainer(GetRouteManager);
+            IDependencyContainerResolver container = DependencyContainerBuilder.SetUpApiContainer(GetRouteManager, GetConnectionString);
             config.DependencyResolver = (IDependencyResolver) container.Resolve(typeof(IDependencyResolver));
```
Sorry if I forget about it before, however, adding generic `Resolve` method would make the code more readable and potentially even less error-prone.

PetrSvirak commented on NotesBackEnd/src/Notes.Api/App_Start/DependencyInjectionConfig.cs:
```diff
@@ -11,11 +12,14 @@ internal static class DependencyInjectionConfig
         private static IRouteManager GetRouteManager()
             => new RouteManager();
+
+        private static IConnectionStringManager GetConnectionString()
```
`GetConnectionString` method that returns `DatabaseConnectionManager` is weird. The problem here is naming. See [DatabaseConnectionManager class](https://github.com/kubmir/kentico-onboarding-cs/pull/4/files#diff-a36b1d921db16c90c27c27866b0b6445R6) for further insights then rename the method, please.

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Database/DatabaseConnectionManager.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Notes.Api.Database
+{
+    internal class DatabaseConnectionManager : IConnectionStringManager
```
This implementation belongs to the `Notes.Api.Services` assembly - it will read the `web.config` just fine 😉

PetrSvirak replied:
Avoid `Manager` or `Helper` or `Provider` whenever you can. What you should be trying to provide to your application is some kind of configuration object. Read more on [`Options` pattern](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options) in .NET Core, please - i would vote for `ConnectionOptions` or `ConnectionsConfig` since this seems to be general purpose of this class.

PetrSvirak replied:
Same applies to the `RouteManager` – `RoutesConfig` or `Rout[e|ing]Options` might be a bit more precise…

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Database/DatabaseConnectionManager.cs:
```diff
@@ -0,0 +1,16 @@
+        private readonly string _mongoDatabaseConnectionString;
+
+        public DatabaseConnectionManager()
+            => _mongoDatabaseConnectionString = WebConfigurationManager.ConnectionStrings["MongoDb_Notes_Connection"].ConnectionString;
```
Make the magical string into a well-named string constant

PetrSvirak replied:
https://imgflip.com/i/26a6a5 🤣

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Database/DatabaseConnectionManager.cs:
```diff
@@ -0,0 +1,16 @@
+            => _mongoDatabaseConnectionString = WebConfigurationManager.ConnectionStrings["MongoDb_Notes_Connection"].ConnectionString;
+
+        public string GetDatabaseConnectionString()
+            => _mongoDatabaseConnectionString;
```
Since you return mongo as (not-more-closely-specified) connection string, it is a question whether the name `mongo` shouldn't remain in the connection strings' entry name only. This is just a tiny detail to consider for a few seconds.

PetrSvirak commented on NotesBackEnd/src/Notes.Dependency/Containers/DependencyContainer.cs:
```diff
@@ -21,20 +21,21 @@ private DependencyContainer(IUnityContainer container)
             _unityContainer = container;
```
![image](https://user-images.githubusercontent.com/15132110/37336341-5ef4f572-26b1-11e8-843c-90f2863144e5.png) 😛

PetrSvirak replied:
And why is the class public 😜

PetrSvirak commented on NotesBackEnd/src/Notes.Dependency/ContainersBuilders/DependencyContainerBuilder.cs:
```diff
@@ -4,32 +4,32 @@
     public static class DependencyContainerBuilder
     {
-        public static IDependencyContainerResolver SetUpApiContainer(Func<IRouteManager> getRouteManager)
+        public static IDependencyContainerResolver SetUpApiContainer(Func<IRouteManager> getRouteManager, Func<IConnectionStringManager> getConnectionString)
```
You actually have the `Notes.Api.Services` to avoid extending the method whenever you can. And in this case, you can: simply register the implementation in the `Api.Services` assembly in the usual way once you move it there.

You cannot do so in case of the `RouteManager` since you want its constants to be used in `Controller`s and this cannot be achieved in a better way than this that I know.

PetrSvirak commented on NotesBackEnd/src/Notes.Contracts/Services/Notes/IGetNoteService.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        Task<Note> GetNoteByIdAsync(Guid id);
+
+        Task<Boolean> IsNoteExistingAsync(Guid id);
```
What about `DoesNoteExist`? 😆 
(or, since the method is in a class called `GetNoteService`, `DoesExist` or eventually even `Exists` could do as well)

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -6,55 +6,148 @@
         public async Task<IHttpActionResult> GetAsync(Guid id)
         {
-            var foundNote = await _repository.GetNoteByIdAsync(id);
+            if (!await _getNoteService.IsNoteExistingAsync(id))
```
Hm, do you really need to send `Guid.Empty` down the aisle? I believe that it is a badly formed request that mustn't make it to the database - or - it is part of business logic that should be handled in a special way at least by your service layer…

Before fixing this issue DO start with a test 😉

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -6,55 +6,148 @@
+
+        private void IsNoteTextValid(Note note)
+        {
+            if (note.Text.Trim() == String.Empty)
```
What about `String.IsEmptyOrWritespace`? 😆

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -6,55 +6,148 @@
+
+            if (!await _getNoteService.IsNoteExistingAsync(id))
+            {
+                var addedNote = await _addNoteService.CreateNoteAsync(noteToUpdate);
```
I believe You should be calling [`PostAsync` method](https://github.com/kubmir/kentico-onboarding-cs/pull/4/files#diff-c1453c69896372d056bbef92e1440973R51) here, to ensure that item passes through the same validation process (so one cannot create items using `PutAsync` that wouldn't be created via `PostAsync`)

Do start with a test 😉

PetrSvirak replied:
The side-effect is that `Id` has to be `Guid.Empty` in order for the `Post` method to work with `noteToUpdate`. This raised the following evil thought of mine: 
1. id != noteToUpdate → `409` (see [comment in the validation method](https://github.com/kubmir/kentico-onboarding-cs/pull/4#discussion-diff-174101786R139))
1. noteToUpdate has invalid text, has updated date or has created date → `400`
1. id == Guid.Empty → `201`
1. id not found → `404`
1. id found → `200`

What do you think?
If you like the flow, definitely do start with the tests.

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -6,55 +6,148 @@
+            }
+
+            IsNoteTextValid(note);
+        }
```
You should ensure that both `note.Id` and `id` provided to the `PutAsync` method via URL are same. Imagine a request to update an item with id `-000001` (URL) with a body that contains the item with id `-00002`. I do believe it is not possible to decide which item should be updated in the end (or what was sender's original intention, to be precise).

Do start with a test 😉

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -6,55 +6,148 @@
+            if (note == null)
+            {
+                NoteIsNotValid(nameof(note),
+                    "Note cannot be null!");
```
One really should not let R# decide when to wrap a line for them; or at least set the R# "properly" 😛 
I suggest setting R#'s line length to some relatively high number and wrap long lines manually where fitting. If there are too many or too long arguments in a method call, I would expect each parameter on its own line or separated into a local variable/const/what-ever.

PetrSvirak commented on NotesBackEnd/src/Notes.Services/Notes/UpdateNoteService.cs:
```diff
@@ -0,0 +1,31 @@
+        public async Task<Note> UpdateNoteAsync(Guid updateNoteId, Note note)
+        {
+            var updateTime = _dateService.GetCurrentDateTime();
+            Note noteToUpdate = new Note { Id = updateNoteId, Text = note.Text, CreationDate = note.CreationDate, LastModificationDate = updateTime };
```
When creating a new instance, the type is written right behind `= new` and it is usually considered redundant for the type to appear at the beginning of the line as well. If you used a factory method to create the instance, the type might add some information, but you don't.

PetrSvirak replied:
Be generous and give each property's assignment a line of its own, please.

PetrSvirak replied:
You mustn't rely on `note.CreationDate` since it is a value provided by the user. None of the date time properties would make much sense if it was not server-controlled, your controller should only allow updating the `note.Text`.

You should use the original item from the DB, merge it with data obtained from the user and then save it again – and yes, you should cache the item, so the item is requested from the DB only once and yes you should reuse the service you already have for caching database item 😉 

If you wanted to reorganize your services, be my guest. But be careful not to introduce uncertainty to your service layer - for example, accepting two arguments of the same type in a method where changing the order of arguments changes the outcome is a quite usual breach of certainty…

PetrSvirak commented on NotesBackEnd/src/Notes.Services/Notes/GetNoteService.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public GetNoteService(INotesRepository repository)
+        {
+            _cachedNote = null;
```
It is usually considered redundant to initialize references with `null` - I guess you mention it to explicitly state there is nothing at the beginning of an instance lifecycle, it just is not a customary way in Cloud; that's all.

PetrSvirak commented on NotesBackEnd/src/Notes.Services/Notes/GetNoteService.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Notes.Services.Notes
+{
+    public class GetNoteService : IGetNoteService
```
`internal` 😉

PetrSvirak commented on NotesBackEnd/src/Notes.Services/Notes/AddNoteService.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Notes.Services.Notes
+{
+    public class AddNoteService : IAddNoteService
```
`internal` 😛

PetrSvirak commented on NotesBackEnd/src/Notes.Services/Notes/UpdateNoteService.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Notes.Services.Notes
+{
+    public class UpdateNoteService : IUpdateNoteService
```
`internal` 😩

PetrSvirak commented on NotesBackEnd/src/Notes.Dependency/ContainersBuilders/DependencyContainerBuilder.cs:
```diff
@@ -4,32 +4,32 @@
-            var bootstrapper = new T();
-            return bootstrapper.RegisterType(container);
-        }
+        private static IDependencyContainerRegister RegisterDependency<TBootstrapper>(this IDependencyContainerRegister container) 
```
I'm still wrestling with container-related names a bit… They seem to be a bit too long and clumsy. What about having, for example, `IResolver` and `IContainer`? You could go with `IDependencyResolver` and/or `IDependencyRegister`, however, the former might mislead you on which one originate in what assembly and the later does not sound natural. But, don't be afraid to come up with any other names You might find descriptive and short enough…

PetrSvirak commented on NotesBackEnd/src/Notes.Services/Utils/DateService.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using Notes.Contracts.Services.Utils;
+
+namespace Notes.Services.Utils
```
`Utils` is some JavaScript crappy name, isn't it 😝 
You should always try to find a better name for a class or namespace in C#. Here, for example, you have some classes that wrap static calls - would `StaticWrappers` do? Feel free to find your own significant characteristic for these two classes and name their common folder by this common feature you connect them with.

PetrSvirak commented on NotesBackEnd/src/Notes.Services/Notes/AddNoteService.cs:
```diff
@@ -0,0 +1,33 @@
+        {
+            var dateTime = _dateService.GetCurrentDateTime();
+
+            note.CreationDate = dateTime;
```
I would actually expect the method not to have side effects - since you pass a `note` in and can `await a note` on "the other side" of the function, I expect the `note` I passed in to remain unchanged…

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -68,14 +75,23 @@ public void Init()
         }
 
+        [Test]
+        public async Task GetAsync_NonExistingId()
```
You described tested method ✔️, described what parameters it got ✔️ but did not describe expected outcome ❌  (fix missing parameter or outcome description everywhere 😉)

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -85,6 +101,50 @@ public void Init()
+        [Test]
+        public async Task PostAsync_AddNewNoteWithEmptyText()
+        {
+            var (_, responseMessage) = await GetExecutedResponse<Note>(() => _controller.PostAsync(new Note { Text = "" }));
```
Always use `String.Empty` when you use empty string intentionally, please.

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -85,6 +101,50 @@ public void Init()
+        {
+            var (_, responseMessage) = await GetExecutedResponse<Note>(() => _controller.PostAsync(null));
+
+            Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
```
A "nice to have" feature in all these tests is the occurrence of the invalid field name in the model error…

PetrSvirak commented on NotesBackEnd/tests/Notes.Services.Tests/Notes/AddNoteServiceTests.cs:
```diff
@@ -0,0 +1,77 @@
+
+            var actualNote = await _addService.CreateNoteAsync(NoteDto);
+
+            Assert.That(actualNote, Is.EqualTo(expectedNote));
```
The problem of all your service tests is that you actually do not test them at all.
What you test is NSubstitute - comment all your code from the service except repository call and re-run your tests. They will pass because you actually only test whatever value your mocked repository returns. You must mock the repository in a way to allow `Note`s just pass through.

PetrSvirak commented on NotesBackEnd/src/Notes.Contracts/Services/Utils/IDateService.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Notes.Contracts.Services.Utils
+{
+    public interface IDateService
```
You are missing tests for this one…

PetrSvirak commented on NotesBackEnd/src/Notes.Contracts/Services/Utils/IGuidService.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Notes.Contracts.Services.Utils
+{
+    public interface IGuidService
```
You are missing tets for this one…

PetrSvirak commented on NotesBackEnd/src/Notes.Repository/Repository/NotesRepository.cs:
```diff
@@ -1,34 +1,45 @@
+        {
+            var connectionString = connectionManager.GetDatabaseConnectionString();
+            var mongoClient = new MongoClient(connectionString);
+            var databaseName = new MongoUrl(connectionString).DatabaseName;
```
Actually, the `MongoClient` accepts `MongoUrl` as a constructor parameter as well 😉

PetrSvirak commented on NotesBackEnd/src/Notes.Repository/Repository/NotesRepository.cs:
```diff
@@ -1,34 +1,45 @@
-        public Task<Note> CreateNoteAsync(Note note)
-            => Task.FromResult(Notes[1]);
+        public async Task<IEnumerable<Note>> GetAllNotesAsync()
+            => (await _persistedNotes.FindAsync<Note>(FilterDefinition<Note>.Empty)).ToEnumerable();
```
While expression bodied functions are definitely preferable, not at all costs. Await in braces is harder to read than simply putting the `await` on standalone line and name its result (by assigning it to a well-named variable).

PetrSvirak commented on NotesBackEnd/src/Notes.Contracts/Services/Notes/IGetNoteService.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Notes.Contracts.Services.Notes
+{
+    public interface IGetNoteService
```
This is not a good name - it does not even sound English 😉 
Look the word `get` in a dictionary or find another way to get rid of the noun at the place of an adjective.

PetrSvirak replied:
Same applies to the `IAddNoteService`…

PetrSvirak replied:
Same applies to the `IUpdateNoteService`…

PetrSvirak replied:
I'm starting to re-think whether forbidding the suffix `Service` would actually help. I was always afraid that you (~ any of you) would come with names like `NoteUpdater` or `NewNoteCreator` 😆

kubmir replied:
I made same renamings but I am not sure if it is better now. If not, I will fix it after consultation with you.

PetrSvirak reviewed (changes requested):
There is still quite some work that needs to be done. Most problematic are services' tests and update service itself. I would like you to clear out some things related to the naming and there are also quite some details worth of further attention.

Please, give the code a bit of time and try to look at it as a complete solution - think about people who would (was it a production code) read through the code base, try to enhance their experience and re-read your tests one by one. Mostly, make the code look more consistent and tidy it up whereever you see fit.

The controller tests seem to be growing too complex and it would be best to repeat yourself more in order of making the tests more easily understandable… One typically do this by minimizing common setup methods and moving individual tests' initialization into their `arrange` part (then enventually re-extracting common parts to private methods if applicable).

Attend to the unused project references as well, please. For example, `.Api` has references to like all other `src` projects for no reason and something tells me that `System.Configuration` won't be needed in there soon enough 😉 - revise all references in all assemblies, please.

PetrSvirak commented on NotesBackEnd/src/Notes.Api/ConnectionStrings.config:
```diff
@@ -0,0 +1 @@
+
```
And, please, remove this file from your repository - inability to build or red icon in IDE is good indicator new developer that he is missing something, but having an empty file instead of non-empty is a bit hard to spot…

kubmir replied:
I removed this file but Appveyor build stop working so I added before_build/after_build script to appveyor.yml which creates/deletes ConnectionStrings.config file for build purposes.

Commit:
KA-228 Add generic Resolve method

Commit:
KA-228 Move databaseConnectionManager, solve naming issues with managers

Commit:
KA-228 Refactor ConnectionOptions

Commit:
KA-228 Refactor DependencyContainer

Commit:
KA-228 Refactors names of functions

Commit:
KA-228 Add check for empty id in notes controller's get method

Commit:
KA-228 Add check for inequal ids in put method of note controller

Commit:
KA-228 Refactor services visibility and lambda expressions

Commit:
KA-228 Rename IDependencyContainer[Register/Resolver]

Commit:
KA-228 Rename Utils namespace/folder to StaticWrappers

Commit:
KA-228 Refactor UpdateNoteService

Commit:
KA-228 Refactor NotesControllerTests, check ModelState

Commit:
KA-228 Create new assembly for comparers

In this commit NUnit nuget packages are consolidated to the newest stable version.

Commit:
KA-228 Refactor services tests

Commit:
KA-228 Add tests for IDateService

Commit:
KA-228 Add tests for GuidService

Commit:
KA-228 Refactor NotesRepository

Commit:
KA-228 Refactor AddNoteService

Commit:
KA-228 Refactor names of all services

Commit:
KA-228 Update appveyor config to omit Notes.Comparers

Commit:
KA-228 Remove ConnectionStrings.config file

Commit:
KA-228 Refactor PUT method of notesController

Commit:
KA-228 Remove unused referencies from assemblies

Commit:
KA-228 Refactor controller tests

Commit:
KA-228 Clean solution

PetrSvirak commented on NotesBackEnd/src/Notes.Api/App_Start/DependencyInjectionConfig.cs:
```diff
@@ -11,11 +11,11 @@ internal static class DependencyInjectionConfig
-            IDependencyContainerResolver container = DependencyContainerBuilder.SetUpApiContainer(GetRouteManager);
-            config.DependencyResolver = (IDependencyResolver) container.Resolve(typeof(IDependencyResolver));
+            IResolver container = DependencyContainerBuilder.SetUpApiContainer(GetRouteOptions);
+            config.DependencyResolver = (IDependencyResolver) container.Resolve<IDependencyResolver>();
```
Just in case, right 😛 – this is actually how I initially noticed that your resolve method does not do what it should - you shouldn't need to explicitly change the `type` anywhere it a project this small 😉

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -71,26 +78,34 @@ public NotesController(IUrlLocationHelper locationHelper, IAddNoteService addSer
+            if (id != noteToUpdate.Id)
             {
-                var addedNote = await _addNoteService.CreateNoteAsync(noteToUpdate);
+                return Conflict();
```
That's nice, however, you should (always) provide your user with enough information to [solve the `conflict`](https://tools.ietf.org/html/rfc7231#section-6.5.8). I'm not sure why does not `WebAPI` pursue this concept, but there is at least this way to do so:
````
return Content(
   HttpStatusCode.Conflict,
   $"Identifier provided in URI ({id}) does not match the identifier provided in the body ({noteToUpdate.Id})");
````

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -71,26 +78,34 @@ public NotesController(IUrlLocationHelper locationHelper, IAddNoteService addSer
         public async Task<IHttpActionResult> DeleteAsync(Guid id)
         {
-            if (!await _getNoteService.IsNoteExistingAsync(id))
+            if (!await _retrievalService.Exists(id))
```
What about `Guid.Empty`? I'm sorry if I forget about it before, however, do you want to do a round trip to the database when you know no such `id` exists in it?

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -140,14 +150,21 @@ private void IsNoteValidForUpdate(Note note)
+            }
+        }
+
+        private void IsIdValid(Guid id)
```
It makes perfect sense to name a method that returns a `boolean` something like `IsIdVald`, hower, when the method starts to change the internal state of its class, something more along the `ValidateId` seems to be appropriate (applies to other similar methods as well).

PetrSvirak commented on NotesBackEnd/src/Notes.Contracts/Repository/INotesRepository.cs:
```diff
@@ -7,14 +7,14 @@ namespace Notes.Contracts.Repository
+        Task<Note> DeleteByIdAsync(Guid id);
 
-        Task<Note> UpdateNoteAsync(Guid id, Note noteToUpdate);
+        Task<Note> UpdateAsync(Guid id, Note noteToUpdate);
```
Why do you actually provide an `id` that is part of the `noteToUpdate`? (That's rhetorical since your `UpdateService` suggests you can simply pass `noteToUpdate` and I also don't like when you change this interface implementation in this very task – it seems more useful to show you it is not necessary (and usually wrong) to change your interface in favour of its (only) implementation - it should reflect the rest of your system requirements, really) While it is understandable in the `service` layer, the repository - being a lever lower - should expect the `noteToUpdate` to be correct and complete…

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NoteController/NoteControllerTestsBase.cs:
```diff
@@ -0,0 +1,92 @@
+        protected NotesController Controller;
+
+        [SetUp]
+        public void SetUpBaseClass()
```
These methods are usually called with as precise names as possible (for both understanding their purpose and to avoid potential name collisions when multilayer inheritance kicks in). Something like `[SetUp|Mock]Controller` would be a bit more practical, I think…

PetrSvirak commented on NotesBackEnd/tests/Notes.Services.Tests/Wrappers/GuidWrapperTests.cs:
```diff
@@ -0,0 +1,39 @@
+        [Test]
+        public void GenerateGuid_Generate100Guids_UniqueGuidsReturned()
+        {
+            List<Guid> generatedGuids = new List<Guid>();
```
There can be the `var` here and (type) information will be lost - `var` has constant length, so it surprisingly (at least it was surprising for me) makes the code more readable…

PetrSvirak commented on NotesBackEnd/tests/Notes.Services.Tests/Wrappers/GuidWrapperTests.cs:
```diff
@@ -0,0 +1,39 @@
+            for (int i = 0; i < 100; i++)
+            {
+                generatedGuids.Add(_guidWrapper.GetNew());
+            }
```
This is perfect for this purpose. Nothing needs to be changed here (maybe the `magical 100` could have been a well-named constant, but that is it). If you though happen to be fan of more functional way to write the same, there is an alaternative:
````
const int NUMBER_OF_RANDOM_IDS = 100;
var generatedGuids = Enumerable
    .Repeat<Func<Guid>>(() => _guidWrapper.GetNew(), NUMBER_OF_RANDOM_IDS)
    .Select(generator => generator())
    .ToArray(); // this is necessary here in order to really `act` in the `act` part of the test ;-)
````

PetrSvirak commented on NotesBackEnd/tests/Notes.Services.Tests/Wrappers/DateWrapperTests.cs:
```diff
@@ -0,0 +1,24 @@
+        public void GetCurrentDateTime_CalledTwice_SecondReturnedTimeBiggerThanFirstReturned()
+        {
+            var firstTime = _dateWrapper.GetCurrentDateTime();
+            var secondTime = _dateWrapper.GetCurrentDateTime();
```
Just a note: it might happen that this test will fail on some machines under certain circumstance (then both calls are made during one tick) - if this happened, would you know how to solve it?

kubmir replied:
I think Thread.Sleep(timeout) should solve it - I added this to the code of test method.

PetrSvirak reviewed (changes requested):
Really nice, I believe this is the last review for the task (unless you attend it too harshly). It is now more about consistency and persistence than anything else. It is sometimes a bit boring to make code really look the same and stick to standards you (or worse: me in this case) define, but it is part of the job and it is worth it (at least in long-term).

I miss the `NotesBackEnd.sln.DotSettings` file in the repository (R#'s shared (code style) settings and once you set everything up, you might find `ReSharper → Inspect → Code Issues in Solution` intriguing – for example, my R# now reports [about 30 issues](https://github.com/kubmir/kentico-onboarding-cs/files/1825999/CodeIssuesPerSolution.zip) (both Rado (especially Rado) and Matúš encountered something similar lately).

But despite all that was said above and bellow, I like the code you write very much - it is clear and easy to understand most of the time. You also seem to really think about each individual solution and, for example, your controller tests are a nice showcase for this.

Playing around with R# configuration can take some time - just go through each setting and think about it, set it, see the differences in the preview, persist or revert the setting. All other things should be absolutely fast details (even the one where you need to write a new test 😉). I believe this task to be closed in no time (like tomorrow evening) 😃.

Commit:
KA-228 Apply code styles and formatting

Commit:
KA-228 Add code-style configuration

Commit:
KA-228 Refactor NotesController

Add check for empty guid in DeleteAsync, refactor names of validation methods,
refactor returning of conflict httpActionResult in PutAsync.

Commit:
KA-228 Refactor signature of UpdateAsync in NotesRepository

Commit:
KA-228 Refactor tests
